accord-core/src/main/java/accord/impl/DepsBuilder.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.impl;
+
+import accord.local.SafeCommandStore;
+import accord.local.Status;
+import accord.primitives.*;
+
+import java.util.*;
+import java.util.function.Consumer;
+
+public class DepsBuilder
+{
+    private final TxnId subject;
+    private final Map<Seekable, Builder> builders = new HashMap<>();
+
+    public DepsBuilder(TxnId subject)
+    {
+        this.subject = subject;
+    }
+
+    private static class TxnInfo
+    {
+        private final TxnId txnId;
+        private final Status status;
+        private final Timestamp executeAt;
+        private final List<TxnId> depsIds;
+        private boolean required = false;
+
+        public TxnInfo(TxnId txnId, Status status, Timestamp executeAt, List<TxnId> depsIds)
+        {
+            this.txnId = txnId;
+            this.status = status;
+            this.executeAt = executeAt;
+            this.depsIds = depsIds;
+        }
+    }
+
+    /**
+     * To remove transitive dependencies, we filter out commands that exist in the deps list
+     * We need commands that are not superseded by other commands with later execution times
+     */
+    public static class Builder
```

Comment:
We had discussed offline avoiding this work, instead applying the deletions on the relevant transition of the commands that we are storing here - i.e. on e.g. `PreApplied` deleting any already `PreApplied` dependencies, and marking as deletable any dependencies that are not yet `PreApplied`.

Reply:
> We had discussed offline avoiding this work, instead applying the deletions on the relevant transition of the commands that we are storing here - i.e. on e.g. PreApplied deleting any already PreApplied dependencies, and marking as deletable any dependencies that are not yet PreApplied. Have you had a chance to explore this approach?

I had, yes, but I don’t think we can apply the deletions at the preapplied stage without having cfk data in memory anyway to determine if a txnId can be transitively linked to the more recent deps. As it stands, the happy path doesn’t need to load deps data after the preaccept/accept phases, so pruning eagerly could require more reads in some cases.

> I think maybe this also avoids us needing to maintain a separate FULL deps history? Since we avoid deleting anything that may yet be used by recovery.

I’d imagine the “deps” deps are going to be pretty small. As I understand it, the requirements for pruning the recovery deps and the “deps” deps are different enough that keeping a single set could require loading quite a bit of extra data for all preaccept/accept operations in some query patterns or availability scenarios.

> Once we introduce a new Stable state that represents durably Committed (as discussed with Alexey, Pierre and Fedor) we can trigger on this status rather than PreApplied.

I’m out of the loop on that discussion, but for practical purposes, we can’t _actually_ prune anything until it’s been applied since the CFK listener continues to update the cfks until it is applied (and would quickly unprune operations).

Reply:
> I’d imagine the “deps” deps are going to be pretty small. As I understand it, the requirements for pruning the recovery deps and the “deps” deps are different enough that keeping a single set could require loading quite a bit of extra data for all preaccept/accept operations in some query patterns or availability scenarios.

I don't think so. I think we probably require that they use the same sets. The reason is this: we need the dependencies to be stable before we can prune, else we may falsely erase a dependency relation that isn't durable. Much like the bug report by Alexey et al, if we depend on a `Commit` record that may be overwritten, we may treat a dependency as transitively durable when in fact it isn't. I'm not 100% certain this can cause a problem, but it feels likely. It would take some time to demonstrate it though.

In effect we want to only erase dependencies from the set when we know that the transaction itself is durably committed, and we know that the transaction we adopt it as a dependency from has its _dependencies_ durably committed. This is like the `Stable` property proposed by Alexey et al, but stronger, as I think for a fast-path commit we cannot erase dependencies until we know the commit record has reached a majority (else we may still recover, and need to see these dependencies for the `Accept` - not just for the `BeginRecovery` verb!). So we really need to witness an Apply record in this case for the transaction _via which_ we are treating the dependency as transitively reachable.

At the very least we need to hash out some more this distinction between recovery and standard deps, as I think it is probably unsafe. We can perhaps hop on a call to talk it through?

> I had, yes, but I don’t think we can apply the deletions at the preapplied stage without having cfk data in memory

Can't we just write a deletion record to some `Set` on the CFK, to all be processed on read?

> I’m out of the loop on that discussion, but for practical purposes, we can’t actually prune anything until it’s been applied since the CFK listener continues to update the cfks until it is applied (and would quickly unprune operations).

That's fine really, although I'd say we could almost certainly retain some minimal summary info in each CFK to permit durably erasing most of these (e.g. recording a TxnId below which we do not need dependencies). Or we could unregister the CFK listener before erasing.

Reply:
Hmm, so thinking about this some more, I think we need to be more careful about removal of deps. Consider the following, with transactions t1, t0.2 and t3 that _will_ execute in this order. But where t0.2 is proposed to execute at t0, but eventually executes at t2.

t1 reaches a quorum, and is committed at a quorum.
t0.2 reaches a quorum on preaccept, but its timestamp is rejected so it initiates an Accept round that needs to witness t1
t3 reaches a quorum, and is committed at a quorum. t1 is erased from the deps list.
t0.2 performs its accept round, and _does not witness t1_.

So, I think we require that every dependency t3 has witnessed is committed before we can use t3 to erase any of its dependencies. Since t3 will have witnessed t2, and t2 executes before t3 (but after t1), so by waiting for all of them to commit it is safe to remove them as dependencies.

There is a question of whether we need only the dependent transaction to be stable (durably committed with dependencies), or the dependency as well. I need to think about that some more.

Another issue occurs to me around the current approach of processing the dependency graph to remove transactions, which is that we might orphan some. If we remove dependencies that themselves were the only strong link, we lose the ability to prune them (at least until truncation kicks in).

Say, A<-B<-C, and B and C both become stable, so that B is erased as a dependency. Now we won’t process A’s removal when it becomes stable, as B is no longer present, and C does not reference it. This might occur on a replica that wasn’t involved with selecting C’s dependencies, when other replicas might have legitimately pruned A as a dependency, since we know that B is a durable link to it.